### PR TITLE
First pass at adding TCF 2.0 handling

### PR DESCRIPTION
**Pull request by hhhjort**
The test cases are now in, this can be reviewed.

### CODE REVIEW

File: consentconstants/purposes.go [hunk 1]
```diff
@@ -39,3 +40,87 @@ const (
...
+	// To do basic ad selection vendors can:
+	//   * Use real-time information about the context in which the ad will be shown, to show the ad, including information about the content and
+	//     the device, such as: device type and capabilities, user agent, URL, IP address
+	//   * Use a user's non-precise geolocation data
+	//   * Control the frequency of ads shown to a user.\n* Sequence the order in which ads are shown to a user.
+	//   * Prevent an ad from serving in an unsuitable editorial (brand-unsafe) context
+	// Vendors cannot:
+	//   * Create a personalised ads profile using this information for the selection of future ads.
+	//   * N.B. Non-precise means only an approximate location involving at least a radius of 500 meters is permitted.
+	BasicAdserving Purpose = 2
```
Comment (SyntaxNode) [inline consentconstants/purposes.go]: Does it make sense to name these something like `TCF2BasicAdserving`?

File: consentconstants/purposes.go [hunk 2]
```diff
@@ -39,3 +40,87 @@ const (
 	// for you in other contexts, i.e. on other service, such as websites or apps, over time.
 	Measurement Purpose = 5
 )
+
+// TCF 2.0 Purposes
+const (
+	// Cookies, device identifiers, or other information can be stored or accessed on your device for the purposes presented to you.
```
Comment (SyntaxNode) [inline consentconstants/purposes.go]: Are the "vendors can" and "vendors cannot" copied from the IAB spec?

File: vendorconsent/bitfield20.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package vendorconsent
+
+import "fmt"
+
+func parseBitField20(data consentMetadata20) (*consentBitField20, error) {
```
Comment (SyntaxNode) [inline vendorconsent/bitfield20.go]: Naming is a bit confusing here, IMHO. I initially thought this was parsing bit field 20 instead of for TCF 2.0. Consider putting these into something like a `vendorconsent/tcf2` package.

File: vendorconsent/bitfield20_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	assertUInt16sEqual(t, 14, consent.VendorListVersion())
+	assertUInt16sEqual(t, 10, consent.MaxVendorID())
+
+	purposesAllowed := buildMap(1, 2, 3, 5, 6, 7, 9, 12, 13, 15, 17, 19, 20, 23, 24)
+	for i := uint8(1); i <= 24; i++ {
+		_, ok := purposesAllowed[uint(i)]
+		assertBoolsEqual(t, ok, consent.PurposeAllowed(consentconstants.Purpose(i)))
+	}
+
+	vendorsWithConsent := buildMap(1, 2, 4, 7, 9, 10)
```
Comment (SyntaxNode) [inline vendorconsent/bitfield20_test.go]: If we're focusing on the test for vendor bit fields, the rest of the assertions are not needed IMHO.

File: vendorconsent/consent.go [hunk 5]
```diff
@@ -63,10 +64,15 @@ type VendorConsents interface {
 
 // ParseString parses a Raw (unpadded) base64 URL encoded string.
 func ParseString(consent string) (VendorConsents, error) {
-	decoded, err := base64.RawURLEncoding.DecodeString(consent)
+	pieces := strings.Split(consent, ".")
+	decoded, err := base64.RawURLEncoding.DecodeString(pieces[0])
```
Comment (SyntaxNode) [inline vendorconsent/consent.go]: Do we not support or do we not need the disclosed vendors, allowed vendors, or publisher tc sections?

File: vendorconsent/metadata20.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+// to make sure that functions on it don't overflow the bounds of the byte array.
+type consentMetadata20 []byte
+
+func (c consentMetadata20) Version() uint8 {
+	// Stored in bits 0-5
+	return uint8(c[0] >> 2)
+}
+
+func (c consentMetadata20) Created() time.Time {
+	_ = c[5]
```
Comment (SyntaxNode) [inline vendorconsent/metadata20.go]: Why is this line here?

File: consentconstants/purposes.go (hunk 1 above)
Comment (hhhjort) [reply consentconstants/purposes.go]: IAB wants to deprecate TCF1 as fast as possible, so we should be able to pull all the old names fairly quickly.

File: consentconstants/purposes.go (hunk 2 above)
Comment (hhhjort) [reply consentconstants/purposes.go]: Yes they are ... specifically the definitions found in the GVL

File: vendorconsent/bitfield20.go (hunk 3 above)
Comment (hhhjort) [reply vendorconsent/bitfield20.go]: That would create 3 packages, a tcf1 package, a tcf2 package, and a general package to roll them together. It is doable, but a chuck or work.

File: vendorconsent/consent.go (hunk 5 above)
Comment (hhhjort) [reply vendorconsent/consent.go]: I don't think we need them, and we definitely don't need them for pass 1. If needed, we can modify the TCF 2.0 object creator to take the extra pieces.

File: vendorconsent/metadata20.go (hunk 6 above)
Comment (hhhjort) [reply vendorconsent/metadata20.go]: That is a good question. Its in the TFC 1 version but should be safe to remove.

File: consentconstants/purposes.go [hunk 7]
```diff
@@ -3,6 +3,7 @@ package consentconstants
 // Purpose is one of the IAB GDPR purposes. These appear in:
 //   1. `root.purposes[i]` of the vendor list: https://vendorlist.consensu.org/vendorlist.json
 //   2. PurposesAllowed of the Consent string: https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/Consent%20string%20and%20vendor%20list%20formats%20v1.1%20Final.md#vendor-consent-string-format-
+// NOTE: These only apply to TCF 1.0, and not TCF 2.0
```
[the remark this answers is not in the document]
Comment (hhhjort) [reply consentconstants/purposes.go]: Yes, Initially I was thinking of moving away from the `Purpose` type, but ended up keeping it.

File: vendorconsent/bitfield20_test.go (hunk 4 above)
Comment (hhhjort) [reply vendorconsent/bitfield20_test.go]: I figure it was better to have them than not, and I was basically just duplicating all the tests I found in the TCF 1.0 code.

File: consent2/bitfield.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
+package consent2
+
+import (
+	"fmt"
+)
+
+func parseBitField(data consentMetadata) (*consentBitField, error) {
+	vendorBitsRequired := data.MaxVendorID()
+
+	// BitFields start at bit 213. This means the last three bits of byte 26 are part of the bitfield.
```
Comment (SyntaxNode) [inline consent2/bitfield.go]: The comment is wrong. The bit field starts at bit 230 so it's the last 3 bits of byte 28 we care about. The code looks to be correct.

File: consent2/bitfield.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	// Careful here... vendor IDs start at index 1...
+	if id <= 3 {
+		return byteToBool(f.firstTwo & (0x04 >> id))
+	}
+	return isSet(f.others, uint(id-3))
+}
+
+// byteToBool returns false if val is 0, and true otherwise
+func byteToBool(val byte) bool {
+	if val == 0 {
```
Comment (SyntaxNode) [inline consent2/bitfield.go]: This could just be `val != 0`

File: consent2/metadata.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	"fmt"
+	"time"
+
+	"github.com/prebid/go-gdpr/consentconstants"
+)
+
+// Parse the metadata from the consent string.
+// This returns an error if the input is too short to answer questions about that data.
+func parseMetadata(data []byte) (consentMetadata, error) {
+	if len(data) < 30 {
```
Comment (SyntaxNode) [inline consent2/metadata.go]: Nitpick: I think 29 bytes should be used as the minimum here. The highest bit needed to parse this section is 228.

File: consent2/metadata.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+func (c consentMetadata) MaxVendorID() uint16 {
+	// The max vendor ID is stored in bits 213 - 228 [00000xxx xxxxxxxx xxxxx000]
+	leftByte := byte((c[26]&0x07)<<5 + (c[27]&0xf8)>>3)
+	rightByte := byte((c[27]&0x07)<<5 + (c[28]&0xf8)>>3)
+	return binary.BigEndian.Uint16([]byte{leftByte, rightByte})
+}
+
+func (c consentMetadata) PurposeAllowed(id consentconstants.Purpose) bool {
+	// Purposes are stored in bits 152 - 175. The interface contract only defines behavior for ints in the range [1, 24]...
```
Comment (SyntaxNode) [inline consent2/metadata.go]: We might want to add a check that Purpose < 24, since it's really just an int without any explicit range enforcement. This would need to be an intentional override of this library, but could result in an overflow.

File: api/vendorlist.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-package vendorlist
+package api
```
[the remark this answers is not in the document]
Comment (hhhjort) [reply api/vendorlist.go]: That won't work. I hit a circular dependence when splitting things out, which is why I made the `api` package. There are things in  `venderconsent` that needs to import, and of course `vendorconsent` needs to import `tcf2`.

File: consent2/metadata.go (hunk 10 above)
Comment (hhhjort) [reply consent2/metadata.go]: I think I made it 30 because there is a second bitfield we don't read yet, so practically it is not possible to have a 29 byte consent string if I counted right.

File: vendorconsent/consent20_test.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
+package vendorconsent
```
Comment (SyntaxNode) [inline vendorconsent/consent20_test.go]: Ah, guess to need to leave the '20' in this file name. Makes sense.

File: consentconstants/purposes.go [hunk 14]
```diff
@@ -39,3 +40,87 @@ const (
 	// for you in other contexts, i.e. on other service, such as websites or apps, over time.
 	Measurement Purpose = 5
 )
+
+// TCF 2.0 Purposes:
+const (
```
Comment (SyntaxNode) [inline consentconstants/purposes.go]: Should these go in /consentconstants/tcf2/ to match the vendor consent approach?

File: api/consent.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+	// The VendorListVersion which is needed to interpret this consent string.
+	//
+	// The IAB is hosting these on their webpage. For example, version 2 of the
+	// Vendor List can be found at https://vendorlist.consensu.org/v-2/vendorlist.json
+	//
+	// For other versions, just replace the "v-*" path with the value returned here.
+	// The latest version can always be found at https://vendorlist.consensu.org/vendorlist.json
+	VendorListVersion() uint16
+
```
Comment (guscarreon) [inline api/consent.go]: I see `api/consent.go` declares a number of fields that are consistent with the elements of the consent string elements listed in:
https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/Consent%20string%20and%20vendor%20list%20formats%20v1.1%20Final.md#vendor-consent-string-format-
            
Which, from the URL I infer is stil version `1.1` isn't it? I'm probably wonrg, but don't we need to update this fields to the TCF 2 consent string listed in this URL instead?
https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20Consent%20string%20and%20vendor%20list%20formats%20v2.md#the-core-string

Out of the consent strings listed in tables found in both URL's the fields are almost identical up to a certain point. In particular, TCF 2 seems to have different fields from `VendorListVersion() uint16` on.

File: consentconstants/tcf2/purposes.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
+package consentconstants
+
+import base "github.com/prebid/go-gdpr/consentconstants"
+
+// TCF 2.0 Purposes:
```
Comment (guscarreon) [inline consentconstants/tcf2/purposes.go]: All the "Regular" purposes found in https://iabeurope.eu/iab-europe-transparency-consent-framework-policies/ are listed. But I also noticed there's a part B "Special purposes" that lists two special purposes and I don't see them listed anywhere. Do we need them at all?

File: api/consent.go (hunk 15 above)
Comment (hhhjort) [reply api/consent.go]: the tcf1 directory has the code for the tcf1 fields, the tcf2 directory code for tcf2 fields. tcf2 is basically tcf1 plus extras, so first pass is just to extract the fields that exist in both versions. Will likely define a second interface which impliments the first plus extra methods for the tcf2 only fields.

File: consentconstants/tcf2/purposes.go (hunk 16 above)
Comment (hhhjort) [reply consentconstants/tcf2/purposes.go]: We don't need them for this pass. We will need them in the future

File: vendorconsent/tcf2/rangesection_test.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+	assertStringsEqual(t, "EN", consent.ConsentLanguage())
+	assertUInt16sEqual(t, 15, consent.VendorListVersion())
+	assertUInt16sEqual(t, 626, consent.MaxVendorID())
+
+	//  The above encoder doesn't support setting purposes.
+	//	purposesWithConsent := buildMap(1, 3, 5, 6, 7, 10)
+	//	for i := uint8(1); i <= 24; i++ {
+	//		_, ok := purposesWithConsent[uint(i)]
+	// 		assertBoolsEqual(t, ok, consent.PurposeAllowed(consentconstants.Purpose(i)))
+	//	}
```
Comment (guscarreon) [inline vendorconsent/tcf2/rangesection_test.go]: We can probably remove this commented lines

File: vendorconsent/consent_test.go [hunk 18]
```diff
@@ -3,6 +3,8 @@ package vendorconsent
 import (
 	"encoding/base64"
 	"testing"
+
+	tcf1 "github.com/prebid/go-gdpr/vendorconsent/tcf1"
 )
 
 func TestIsSet(t *testing.T) {
```
Comment (guscarreon) [inline vendorconsent/consent_test.go]: Folders `vendorconsent/`, `vendorconsent/tcf1/`, and `vendorconsent/tcf2/` all define their own, identical `isSet` function. Maybe we can put only one definition of it in `vendorconsent/utils` object in a future PR.

File: vendorconsent/consent.go (hunk 5 above)
Comment (peczenyj) [reply vendorconsent/consent.go:15]: hey guys, now we need the disclosed vendors, will you accept a Pull Request to add support on it?

Comment (SyntaxNode) [reply vendorconsent/consent.go:15]: > will you accept a Pull Request to add support on it?

Sure will.
